File: firebase/functions/lib/community/create_announcement.dart
import 'dart:async';

import 'package:firebase_admin_interop/firebase_admin_interop.dart';
import 'package:firebase_functions_interop/firebase_functions_interop.dart';
import '../on_call_function.dart';
import '../utils/email_templates.dart';
import '../utils/firestore_utils.dart';
import '../utils/notifications_utils.dart';
import 'package:data_models/cloud_functions/requests.dart';
import 'package:data_models/community/community_user_settings.dart';
import 'package:data_models/community/membership.dart';

class CreateAnnouncement extends OnCallMethod<CreateAnnouncementRequest> {
  NotificationsUtils notificationsUtils;

  CreateAnnouncement({NotificationsUtils? notificationsUtils})
      : notificationsUtils = notificationsUtils ?? NotificationsUtils(),
        super(
          'sendAnnouncement',
          (jsonMap) => CreateAnnouncementRequest.fromJson(jsonMap),
        );

  @override
  Future<void> action(
    CreateAnnouncementRequest request,
    CallableContext context,
  ) async {
    final communityMembershipDoc = await firestore
        .document(
          'memberships/${context.authUid}/community-membership/${request.communityId}',
        )
        .get();

    final membership = Membership.fromJson(
      firestoreUtils
          .fromFirestoreJson(communityMembershipDoc.data.toMap() ?? {}),
    );

    if (!membership.isAdmin) {
      throw HttpsError(HttpsError.failedPrecondition, 'unauthorized', null);
    }

    final docRef = firestore
        .collection('/community/${request.communityId}/announcements')
        .document();
    final docData = DocumentData.fromMap(
      firestoreUtils.toFirestoreJson(request.announcement!.toJson()),
    );

    await docRef.setData(docData);

    print('Sending notification');
    await notificationsUtils.sendCommunityNotifications(
      communityId: request.communityId,
      filterUsersBy: (settings) =>
          settings.notifyAnnouncements == NotificationEmailType.immediate ||
          settings.notifyAnnouncements == null,
      generateMessage: ({
        required community,
        required user,
        required unsubscribeUrl,
      }) =>
          SendGridEmailMessage(
        subject: 'New Announcement: ${request.announcement?.title}',
        html: makeNewAnnouncementBody(
          community: community,
          announcement: request.announcement,
          unsubscribeUrl: unsubscribeUrl,
        ),
      ),
    );
  }
}

File: firebase/functions/src/community/create_announcement.ts
import * as functions from 'firebase-functions'
import { HttpsError } from 'firebase-functions/lib/providers/https'
import { firestore, firestoreUtils } from '../utils/infra/firestore_utils'
import { OnCallMethod } from '../on_call_function'
import {
    CreateAnnouncementRequest,
    Membership,
    CommunityUserSettings,
    NotificationEmailType,
    SendGridEmailMessage,
} from '../types'
import { NotificationsUtils } from '../utils/notifications_utils'
import { membershipIsAdmin } from '../types'

export class CreateAnnouncement extends OnCallMethod<CreateAnnouncementRequest> {
    notificationsUtils: NotificationsUtils

    constructor(opts?: { notificationsUtils?: NotificationsUtils }) {
        super('sendAnnouncement', (json) => json as CreateAnnouncementRequest)
        this.notificationsUtils = opts?.notificationsUtils ?? new NotificationsUtils()
    }

    async action(
        request: CreateAnnouncementRequest,
        context: functions.https.CallableContext
    ): Promise<void> {
        const membershipDoc = await firestore
            .doc(`memberships/${context.auth?.uid}/community-membership/${request.communityId}`)
            .get()
        const membership = membershipDoc.data() as Membership | undefined

        if (!membership?.status || !membershipIsAdmin(membership.status)) {
            throw new HttpsError('failed-precondition', 'unauthorized')
        }

        const docRef = firestore.collection(`/community/${request.communityId}/announcements`).doc()
        await docRef.set(
            firestoreUtils.toFirestoreJson(
                request.announcement as unknown as Record<string, unknown>
            )
        )

        console.log('Sending notification')
        await this.notificationsUtils.sendCommunityNotifications({
            communityId: request.communityId,
            filterUsersBy: (settings: CommunityUserSettings) =>
                settings.notifyAnnouncements === NotificationEmailType.immediate ||
                settings.notifyAnnouncements == null,
            generateMessage: ({ community, user, unsubscribeUrl }) => ({
                subject: `New Announcement: ${request.announcement?.title}`,
                html: `<p>New announcement in ${community.name}</p><p>Unsubscribe: ${unsubscribeUrl}</p>`,
            }),
        })
    }
}
